### ISSUE TITLE
Make dimensions non mandatory

### ISSUE THREAD

**Issue by celiasmith** [nengo/nengo#1480]
I'm wondering if it might make sense to change the api to make the `dimensions` parameter not mandatory for ensembles, and default it to zero.  This would make things cleaner when defining networks e.g. deepnets where you don't care about the state space dimensions.  I don't think it would negatively impact anything else.

**Comment by tbekolay** [comment]
Makes sense to me! `dimensions==0` is a nice easy case to look for in validation and builder logic so it should be relatively easy to support.

**Comment by tcstewar** [comment]
I'm also in favour of that.  It'd make it more natural to implement non-NEF networks in general, which is something we want to do more of....

**Comment by drasmuss** [comment]
`dimensions=None` might be a bit more semantically clear (I'd read that as "this ensemble doesn't have a state-space representation", as opposed to "this ensemble has a state-space representation with 0 dimensions").  But definitely agreed on making it optional!

**Comment by hunse** [comment]
Would we also not generate encoders for such ensembles, and fail if someone tries to do e.g. an ensemble->ensemble connection with them? Things might look a bit different if we get rid of `Ensemble.neurons` and add `NEFTransform`, too.

**Comment by tbekolay** [comment]
> we also not generate encoders for such ensembles

Yeah, I think that sounds right. With the `NEFTransform` we could move the encoder to the connection, though then we'll have to ensure that all incoming connections share those encoders, which might be tricky API-wise.
